[MRG + 1] Reject regression type targets for classifiers

vermouthmjl opened the pull request:
This PR solves https://github.com/scikit-learn/scikit-learn/issues/5060

vermouthmjl committed:
a commun test to check if classifiers fail when fed regression targets

vermouthmjl committed:
Add non regression target check in classifiers' `fit` method

amueller commented on sklearn/utils/multiclass.py:
```diff
@@ -229,6 +229,26 @@ def is_multilabel(y):
+            'multilabel-indicator', 'multilabel-sequences']:
+        if not y_type is 'continuous':
+            raise ValueError("Unknown label type: %r" % y)
+        elif len(np.unique(y)) > 2:
```
Why that?

vermouthmjl replied:
When y is centered (in a test for RandomizedLogisticRegression), [0,1] becomes [-0.5, 0.5], therefore is continuous.
This is a bit arbitrary... Maybe it's better to change that test.

amueller replied:
That seems like a bug in randomized logistic regression.

amueller commented on sklearn/utils/multiclass.py:
```diff
@@ -229,6 +229,26 @@ def is_multilabel(y):
     """
     return is_label_indicator_matrix(y) or is_sequence_of_sequences(y)
 
+def assert_non_regression_targets(y):
```
Can you test the function?

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1309,3 +1310,11 @@ def fit(self, X, y):
+    boston = load_boston()
+    X, y_ = boston.data, boston.target
+    e = Estimator()
+    assert_raises(ValueError, e.fit, X, y_)
```
Can you check for the error message raised with ``assert_raise_regexp`` or assert_raise_message``?

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1309,3 +1310,11 @@ def fit(self, X, y):
+    # Check if classifier throws an exception when fed regression targets
+
+    boston = load_boston()
+    X, y_ = boston.data, boston.target
```
why ``y_`` ?

amueller commented:
Looks good apart from my minor nitpicks.
I'm wondering if we should include this check in ``check_X_y`` (if y is not supposed to be numeric) but I'm not sure if this is a good idea.

vermouthmjl committed:
test assert_non_regression_targets, no longer considers continuous with two distinct values

vermouthmjl committed:
Labels shouldn't be centered.

vermouthmjl commented:
OK, I added a test for assert_non_regression_targets, and changed the RandomizedLogisticRegression test so that there is no need to treat the 2 float value targets.

mblondel commented on sklearn/ensemble/forest.py:
```diff
@@ -417,6 +418,8 @@ def _set_oob_score(self, X, y):
         self.oob_score_ = oob_score / self.n_outputs_
 
     def _validate_y_class_weight(self, y):
+        assert_non_regression_targets(y)
```
The name `assert_*` makes it sound like a unit test utility function.

amueller replied:
True

vermouthmjl replied:
I thought about is_non_regression_targets, but it is supposed to do nothing at all when it is non regression targets, and raise an exception otherwise. So returning a bool seemed a little superfluous.

amueller replied:
you could do ``check_non_regression_target``? On second though, as it raises an exception, assert seems ok.

jmschrei replied:
`check` functions seem to raise exceptions too, specifically `check_array`. I like `check_non_regression_target` better than `assert_non_regression_target`, personally.

vermouthmjl replied:
OK. I changed that into check_non_regression_targets.

mblondel commented:
One concern is that `type_of_target` does potentially expensive things if n_samples is large:

https://github.com/vermouthmjl/scikit-learn/blob/type_of_target/sklearn/utils/multiclass.py#L328
https://github.com/vermouthmjl/scikit-learn/blob/type_of_target/sklearn/utils/multiclass.py#L330

amueller commented:
The first one is only done when ``y.dtype`` is float. So I'd blame the user ;)
The unique is done again afterwards in the classifier, so it is not really expensive compared to what happens next. We could avoid duplicate work by replicating the float test in the new helper and not calling ``type_of_target``.

amueller commented:
We could add an ``is_regression_target`` and use that in the classifiers and in ``type_of_target`` maybe?

jayflo commented:
vermouthmjl : are you completing [#4976](https://github.com/scikit-learn/scikit-learn/issues/4976) as well?

amueller commented:
What is the status here? Can you please rebase?

amueller commented:
It looks like it is good to go. @ogrisel any opinions?

vermouthmjl committed:
remove is_sequence_of_sequence as in 72f6225f7be682b2667ae397ed5c6fd8979bd94e

vermouthmjl committed:
check non regression targets in QuadraticDiscriminantAnalysis

jmschrei commented on sklearn/utils/multiclass.py:
```diff
@@ -158,6 +158,88 @@ def is_multilabel(y):
 
 
+
+# def is_sequence_of_sequences(y):
```
Why is this commented function added in?

jmschrei commented on sklearn/utils/multiclass.py:
```diff
@@ -158,6 +158,88 @@ def is_multilabel(y):
+#     return out
+
+
+# def is_multilabel(y):
```
same

jmschrei commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1443,3 +1444,11 @@ def fit(self, X, y):
+    boston = load_boston()
+    X, y = boston.data, boston.target
+    e = Estimator()
+    assert_raise_message(ValueError, "Unknown label type", e.fit, X, y)
```
newline needed

vermouthmjl committed:
new line at the end of file, delete commented function

vermouthmjl committed:
change assert_non_regression_targets into check_non_regression_targets

amueller commented:
Hm I just realize we should have  a common test for this. Could you please add one in ``estimator_checks``?

vermouthmjl commented:
@amueller I think I already did that by adding `check_classifiers_regression_target` in `estimator_checks`. Maybe I misunderstood what you mean by common test?

amueller commented:
@vermouthmjl sorry, you're good. I was a bit sleep deprived.
@ogrisel merge?

ogrisel commented on sklearn/linear_model/logistic.py:
```diff
@@ -1123,9 +1124,10 @@ def fit(self, X, y, sample_weight=None):
-
-        X, y = check_X_y(X, y, accept_sparse='csr', dtype=np.float64,
+            
+        X, y = check_X_y(X, y, accept_sparse='csr', dtype=np.float64, 
```
cosmetics: trailing spaces

ogrisel commented on sklearn/utils/multiclass.py:
```diff
@@ -157,6 +157,24 @@ def is_multilabel(y):
+        raise ValueError("Unknown label type: %r" % y)
+
+
 
```
PEP8: only 2 blank lines to separate top level blocks (e.g. class or function definitions).

ogrisel commented on sklearn/utils/multiclass.py:
```diff
@@ -157,6 +157,24 @@ def is_multilabel(y):
                                     _is_integral_float(labels))
 
+def check_non_regression_targets(y):
+    """Determine if target `y` is of non regression type
```
Ensure that target `y` is of a non-regression type.

ogrisel commented on sklearn/utils/multiclass.py:
```diff
@@ -157,6 +157,24 @@ def is_multilabel(y):
+    For the moment, the following (as defined in type_of_target) are allowed:
+        - 'binary', 'multiclass', 'multiclass-multioutput', 
+        'multilabel-indicator', 'multilabel-sequences'
+        - 'continuous', if only there are only two unique values
```
This is not matching the content of the function. I think the content of the function is more correct.

ogrisel commented on sklearn/utils/multiclass.py:
```diff
@@ -157,6 +157,24 @@ def is_multilabel(y):
+    """
+    y_type = type_of_target(y)
+    if y_type not in ['binary', 'multiclass', 'multiclass-multioutput', 
+            'multilabel-indicator', 'multilabel-sequences']:
```
PEP8 indentation issue (I think).

ogrisel commented on sklearn/utils/multiclass.py:
```diff
@@ -157,6 +157,24 @@ def is_multilabel(y):
         return len(labels) < 3 and (y.dtype.kind in 'biu' or  # bool, int, uint
                                     _is_integral_float(labels))
 
+def check_non_regression_targets(y):
```
I think this function would better be named `check_classification_targets` instead. WDYT?

vermouthmjl replied:
sure, that sounds more reasonable

MechCoder commented:
Is it clearer to move the `check_non_regression_targets` to `ClassifierMixin` or does it make the code look more complicated?

ogrisel commented:
> Is it clearer to move the check_non_regression_targets for every fit to ClassifierMixin.fit or does it make the code look more complicated (because one has to call the superclass.fit) every time?

I don't think a Mixin class should implement fit to do input checks without performing any actual data fitting and then expect the concrete classes to override it to do the actual data fitting.

I find it simpler and more explicit to call input check functions explicitly in the fit method of the concrete class.

MechCoder commented:
Makes sense. LGTM as well (apart from inline comments).

Maybe you can rebase and merge yourself?

vermouthmjl committed:
change the name check_non_regression_targets() to check_classification_targets()

vermouthmjl committed:
using assert_raises_regex in common test and test for check_classification_targets

vermouthmjl commented:
I rebased. @amueller

amueller commented:
squashed and merged.  I think it will be good to have this in the RC.
